src/Base/ContractManager.h

```diff
@@ -0,0 +1,113 @@
+ *                                                                          *
+ *   This file is part of FreeCAD.                                          *
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#ifndef CONTRACTMANAGER_H
+#define CONTRACTMANAGER_H
+
+#include <algorithm>
+#include <deque>
+#include <map>
+#include <optional>
+#include <vector>
+#include <list>
+
+namespace Base
+{
+
+class ContractManager
+{
+public:
+    ContractManager() = default;
+
+    template<typename T>
+    T* provide() const
+    {
+        const auto contractName = typeid(T).name();
+
+        if (auto it = _contracts.find(contractName); it != _contracts.end()) {
+            auto contracts = it->second;
+
+            if (contracts.empty()) {
+                return nullptr;
+            }
+
+            return static_cast<T*>(it->second.front());
+        }
+
+        return nullptr;
+    }
+
+    template<typename T>
+    std::list<T*> all() const
+    {
+        if (auto it = _contracts.find(typeid(T).name()); it != _contracts.end()) {
+            auto source = it->second;
+
+            std::list<T*> result(source.size());
+
+            std::transform(source.begin(), source.end(), result.begin(), [](void* ptr) {
+                return static_cast<T*>(ptr);
+            });
+
+            return result;
+        }
+
+        return {};
+    }
+
+    template<typename T>
+    void implement(T* contract)
+    {
+        _contracts[typeid(T).name()].push_front(contract);
+    }
+
+    static ContractManager& get()
+    {
+        if (!_instance) {
+            _instance = new ContractManager();
+        }
+
+        return *_instance;
+    }
+
+private:
+    std::map<const char*, std::deque<void*>> _contracts;
```

Comment:
Using `void*`  isn't so nice. I wonder if type erasure using proper C++ can be used here: https://www.modernescpp.com/index.php/type-erasure/